src/shogun/evaluation/CrossValidatedCalibration.cpp

```diff
@@ -0,0 +1,399 @@
+		    m_splitting_strategy->generate_subset_indices(i);
+
+		/* set subset for testing labels */
+		m_labels->add_subset(subset_indices);
+
+		/* produce output for desired indices */
+		CCalibration* calibrator = new CCalibration();
+		calibrator->set_machine((CMachine*)m_machine->clone());
+		calibrator->set_labels((CLabels*)m_labels->clone());
+		calibrator->set_calibration_method(
+		    (CCalibrationMethod*)m_calibration_method->clone());
+
+		bool trained = calibrator->train_locked(subset_indices);
+
+		if (!trained)
+		{
+			return false;
+		}
+
+		m_calibration_machines->set_element(calibrator, i);
+
+		/* remove subset to prevent side effects */
+		m_labels->remove_subset();
+
+		SG_DEBUG("done locked calibration\n", get_name())
+	}
+	return true;
+}
+
+CLabels* CCrossValidatedCalibration::apply_once(
+    CMachine* machine, SGVector<index_t> subset_indices)
+{
+	return machine->apply_locked(subset_indices);
+}
+
+CLabels*
+CCrossValidatedCalibration::apply_once(CMachine* machine, CFeatures* features)
+{
+	return machine->apply(features);
+}
+
+template <typename T>
+CBinaryLabels* CCrossValidatedCalibration::get_binary_result(T data)
+{
+	index_t num_machines = m_calibration_machines->get_num_elements();
+
+	CBinaryLabels* temp_result;
+	CMachine* temp_machine;
+	SGVector<float64_t> result_values;
+	CBinaryLabels* result_labels;
+	CLabels* result;
+
+	temp_machine = (CMachine*)m_calibration_machines->get_element(0);
+	result = apply_once(temp_machine, data);
+	result_labels = CLabelsFactory::to_binary(result);
+	result_values = result_labels->get_values();
+
+	for (index_t i = 0; i < num_machines; ++i)
+	{
+		temp_machine = (CMachine*)m_calibration_machines->get_element(0);
+		result = apply_once(temp_machine, data);
+		temp_result = CLabelsFactory::to_binary(result);
+		result_values += temp_result->get_values();
+	}
+
+	linalg::scale(result_values, result_values, 1.0 / num_machines);
+	result_labels->set_values(result_values);
+
+	return result_labels;
+}
+
+CBinaryLabels* CCrossValidatedCalibration::apply_binary(CFeatures* features)
+{
+	return get_binary_result(features);
+}
+
+CBinaryLabels* CCrossValidatedCalibration::apply_locked_binary(
+    SGVector<index_t> subset_indices)
+{
+	return get_binary_result(subset_indices);
+}
+
+template <typename T>
+CMulticlassLabels* CCrossValidatedCalibration::get_multiclass_result(T data)
+{
+	index_t num_machines = m_calibration_machines->get_num_elements();
+
+	CMulticlassLabels* labels = CLabelsFactory::to_multiclass(m_labels);
+	index_t num_classes = labels->get_num_classes();
+
+	CMulticlassLabels* temp_result;
+	CMachine* temp_machine;
+	SGVector<float64_t> result_values, temp_values;
+	CMulticlassLabels* result_labels;
+	CLabels* result;
+
+	temp_machine = (CMachine*)m_calibration_machines->get_element(0);
+	result = apply_once(temp_machine, data);
+	result_labels = CLabelsFactory::to_multiclass(result);
+
+	for (index_t i = 1; i < num_machines; ++i)
```

Comment:
why from 1 and not 0?

Reply:
The result from the first machine has already retrieved just before the loop. This is a dirty hack that I used because I couldn't find any method to get the number of samples in the features.

Reply:
I would still start the loop at 0 for readability

Reply:
Understood. However, what is the standard way of getting the number of samples from a given set of feature? (i.e. how do I know how many labels will be there in the predictions before actually calling apply?)

Reply:
You can still get the first element to do that.
And then you should ensure that all of the features have the same number of elements

Reply:
Wait, isnt that just `get_num_vectors' ?